java/arcs/sdk/testing/BaseTestHarness.kt

```diff
@@ -0,0 +1,92 @@
+package arcs.sdk.testing
+
+import arcs.core.data.HandleMode
+import arcs.core.host.EntityHandleManager
+import arcs.core.storage.driver.RamDisk
+import arcs.core.storage.driver.RamDiskDriverProvider
+import arcs.core.storage.handle.HandleManager
+import arcs.core.storage.keys.RamDiskStorageKey
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
+import arcs.jvm.util.JvmTime
+import arcs.sdk.EntitySpec
+import arcs.sdk.Handle
+import arcs.sdk.Particle
+import arcs.sdk.testing.HandleFlavor.COLLECTION
+import arcs.sdk.testing.HandleFlavor.SINGLETON
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+/**
+ * Indicates the flavor of a handle: Singleton or a Collection (more may come).
+ */
+enum class HandleFlavor {
+    SINGLETON,
+    COLLECTION
+}
+
+/**
+ * Groups all the per-handle information that's relevant to the TestHarness.
+ */
+data class HandleDescriptor(
+    val name: String,
+    val entitySpec: EntitySpec<*>,
+    val flavor: HandleFlavor
+)
+
+/**
+ * A base class for Test Harnesses for Kotlin particles.
+ */
+open class BaseTestHarness<P : Particle>(
+    private val factory: () -> P,
+    private val descriptors: List<HandleDescriptor>,
+    scope: CoroutineScope
+) {
+    protected val handles: MutableMap<String, Handle> = mutableMapOf()
+    lateinit var particle: P
+
+    init {
+        RamDisk.clear()
+        RamDiskDriverProvider()
+        RamDiskStorageKey.registerParser()
+        RamDiskStorageKey.registerKeyCreator()
+
+        scope.launch {
+            val handleManager = EntityHandleManager(HandleManager(JvmTime))
+            descriptors.forEach {
+                val storageKey = ReferenceModeStorageKey(
+                    backingKey = RamDiskStorageKey("backing_${it.name}"),
+                    storageKey = RamDiskStorageKey("entity_${it.name}")
+                )
+
+                val handle = when (it.flavor) {
+                    SINGLETON -> handleManager.createSingletonHandle(
+                        HandleMode.ReadWrite,
+                        it.name,
+                        it.entitySpec,
+                        storageKey,
+                        it.entitySpec.SCHEMA)
+                    COLLECTION -> handleManager.createCollectionHandle(
+                        HandleMode.ReadWrite,
+                        it.name,
+                        it.entitySpec,
+                        storageKey,
+                        it.entitySpec.SCHEMA)
+                }
+
+                handles[it.name] = handle
+            }
+        }
+    }
+
+    /**
+     * Creates a particle and plays its boot up sequence.
```

Comment:
Question to answer with KDoc: "what is a particle's boot up sequence?"

Reply:
This is good question in general. I hear it quite often... Added a TODO to update the documentation once we settle on all the lifecycle methods.